sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/UnboundedSourceImpl.java

```diff
@@ -0,0 +1,121 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.pubsublite.internal;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkState;
+
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.PartitionLookupUtils;
+import com.google.cloud.pubsublite.proto.SequencedMessage;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.extensions.protobuf.ProtoCoder;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.io.gcp.pubsublite.SubscriberOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Optional;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+public class UnboundedSourceImpl extends UnboundedSource<SequencedMessage, CheckpointMarkImpl> {
+
+  private final SubscriberOptions subscriberOptions;
+  private final SubscriberFactory subscriberFactory;
+  private final BacklogReaderFactory readerFactory;
+  private final Optional<Partition> partition;
+
+  interface SubscriberFactory extends Serializable {
+
+    MemoryBufferedSubscriber create(SubscriptionPartition subscriptionPartition, Offset offset);
+  }
+
+  interface BacklogReaderFactory extends Serializable {
+
+    TopicBacklogReader create(SubscriptionPartition subscriptionPartition);
+  }
+
+  UnboundedSourceImpl(
+      SubscriberOptions subscriberOptions,
+      SubscriberFactory subscriberFactory,
+      BacklogReaderFactory readerFactory) {
+    this.subscriberOptions = subscriberOptions;
+    this.subscriberFactory = subscriberFactory;
+    this.readerFactory = readerFactory;
+    this.partition = Optional.absent();
+  }
+
+  private UnboundedSourceImpl(
+      SubscriberOptions subscriberOptions,
+      SubscriberFactory subscriberFactory,
+      BacklogReaderFactory readerFactory,
+      Partition partition) {
+    this.subscriberOptions = subscriberOptions;
+    this.subscriberFactory = subscriberFactory;
+    this.readerFactory = readerFactory;
+    this.partition = Optional.of(partition);
+  }
+
+  @Override
+  public List<? extends UnboundedSource<SequencedMessage, CheckpointMarkImpl>> split(
+      int desiredNumSplits, PipelineOptions options) throws Exception {
+    checkState(!partition.isPresent());
+    int numPartitions = PartitionLookupUtils.numPartitions(subscriberOptions.subscriptionPath());
```

Comment:
Is there ever a crazy # of partitions? Would we want multiple partitions to a single source shard.  I believe Kafka did this.  I think one concern is per-shard information building up in the graph submitted to dataflow.

Reply:
Not really- that would require the user to pay a crazy amount of money.

per:
>  Would we want multiple partitions to a single source shard

This adds a lot of complexity to the watermarking and commit process, and the low performance risk is mitigated by minimum spends on a per-partition basis in Pub/Sub Lite.

If this ever becomes an issue for a customer, they will be using enough Pub/Sub Lite resources that I will personally go in and add the complexity.